datanode/topology/dynamic.go

```diff
@@ -0,0 +1,242 @@
+package topology
+
+import (
+	"errors"
+	"github.com/m3db/m3/src/cluster/placement"
+	"github.com/m3db/m3/src/cluster/services"
+	"github.com/m3db/m3/src/cluster/shard"
+	xwatch "github.com/m3db/m3/src/x/watch"
+	"github.com/uber/aresdb/common"
+	aresShard "github.com/uber/aresdb/datanode/shard"
+	"github.com/uber/aresdb/utils"
+	"go.uber.org/zap"
+	"sync"
+)
+
+var (
+	errInvalidService            = errors.New("service topology is invalid")
+	errUnexpectedShard           = errors.New("shard is unexpected")
+	errMissingShard              = errors.New("shard is missing")
+	errNotEnoughReplicasForShard = errors.New("replicas of shard is less than expected")
+	errInvalidTopology           = errors.New("could not parse latest value from config service")
+)
+
+type dynamicInitializer struct {
+	sync.Mutex
+
+	opts DynamicOptions
+	topo Topology
+}
+
+// NewDynamicInitializer returns a dynamic topology initializer
+func NewDynamicInitializer(opts DynamicOptions) Initializer {
+	return &dynamicInitializer{opts: opts}
+}
+
+func (i *dynamicInitializer) Init() (Topology, error) {
+	i.Lock()
+	defer i.Unlock()
+
+	if i.topo != nil {
+		return i.topo, nil
+	}
+
+	topo, err := newDynamicTopology(i.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	i.topo = topo
+	return i.topo, nil
+}
+
+func (i *dynamicInitializer) TopologyIsSet() (bool, error) {
+	return true, nil
```

Comment:
nit this function is dummy.. should this be like `return i.topo != nil, nil` at least?